framework/aster-frame/src/sync/wait.rs

```diff
@@ -136,13 +164,16 @@ impl WaitQueue {
     }
 
     fn is_empty(&self) -> bool {
-        self.num_wakers.load(Ordering::Acquire) == 0
+        // On x86-64, this generates `mfence; mov`, which is exactly the right way to implement
+        // atomic loading with `Ordering::Release`. It performs much better than naively
+        // translating `fetch_add(0)` to `lock; xadd`.
+        self.num_wakers.fetch_add(0, Ordering::Release) == 0
     }
 
     fn enqueue(&self, waker: Arc<Waker>) {
         let mut wakers = self.wakers.lock_irq_disabled();
         wakers.push_back(waker);
-        self.num_wakers.fetch_add(1, Ordering::Release);
+        self.num_wakers.fetch_add(1, Ordering::Acquire);
```

Comment:
The `fetch_add` is already within a lock-based critical section. So I think `Relaxed` would be enough.

Reply:
> As soon as the waiter is woken up by the waker, it must see the true condition. This is trivially satisfied if `wake_up()` and `wait()` synchronize with a lock.

To "synchronize with a lock", it is necessary that _both_ sides held the lock. However, the `is_empty()` fast path does not hold any lock. So the synchronization with lock may never occur.

In this case, it is the `Ordering::Release` in `is_empty()` synchronizes with this `Ordering::Acquire`, so it must be a `Ordering::Acquire`, otherwise the synchronization does not work, either.

From the CPU's point of view, an `Ordering::Acquire` ensures that the reads and writes _after_ the load are not _ordered before_ the load. If this line uses `Ordering::Relaxed`, the condition test _after_ `enqueue` will have a chance to be _reordered before_ the `num_waker` update, which will cause problems.

The lock release (which happens after the `num_waker` update but before the condition test) should use `Ordering::Release`, which _can_ ensure that the statements before the lock release cannot be _ordered after_ the lock release, but _cannot_ prevent the condition test after the lock release from being _reordered before_.